internal/operators/operators.go

```diff
@@ -0,0 +1,301 @@
+	Context() context.Context
+	Operators() Operators
+	Logger() logger.Logger
+}
+
+type (
+	EnricherFunc func(any) error
+)
+
+var ErrAsync = errors.New("async")
+
+type Operator interface {
+	// Name must return a unique name for the operator
+	Name() string
+
+	// Description is an optional description to show to the user
+	Description() string
+
+	// Params will return global params (required) for this operator
+	Params() params.ParamDescs
+
+	// PerGadgetParams will return params (required) per gadget instance of the operator
+	PerGadgetParams() params.ParamDescs
+
+	// Dependencies can list other allOperators that this operator depends on
+	Dependencies() []string
+
+	// CanOperateOn should test whether the operator supports the given gadget. Init has not
+	// necessarily been called at this point.
+	CanOperateOn(gadgets.Gadget) bool
+
+	// Init allows the operator to initialize itself
+	Init(params *params.Params) error
+
+	// Close allows the operator to clean up stuff prior to exiting
+	Close() error
+
+	// Instantiate is called before a gadget is run (before PreGadgetRun) with this operator
+	// This must return something that implements operator as well.
+	// This is useful to create a context for an operator by wrapping it.
+	// Params given here are the ones returned by PerGadgetParams()
+	Instantiate(runner Runner, gadgetInstance any, perGadgetParams *params.Params) (OperatorInstance, error)
+}
+
+type OperatorInstance interface {
+	// PreGadgetRun is called before a gadget is started but after all allOperators have been initialized
+	PreGadgetRun() error
+
+	// PostGadgetRun is called on the operator that was returned from PrepareTrace after a
+	// gadget was stopped
+	PostGadgetRun() error
+
+	// Enricher should return a function that is able to operator on the event ev and
+	// call next afterwards.
+	Enricher(next EnricherFunc) EnricherFunc
+
+	// Deprecated: EnrichEvent
+	EnrichEvent(ev any) error
+}
+
+type operatorInterfaces struct {
+	Operator
+	OperatorInstance
+}
+
+type Operators []operatorInterfaces
+
+// KubernetesFromMountNSID is a typical kubernetes operator interface that adds node, pod, namespace and container
+// information given the MountNSID
+type KubernetesFromMountNSID interface {
+	ContainerInfoSetters
+	GetMountNSID() uint64
+}
+
+type ContainerInfoSetters interface {
+	SetContainerInfo(pod, namespace, container string)
+	SetNode(string)
+}
+
+var allOperators = map[string]Operator{}
+
+type operatorWrapper struct {
+	Operator
+	initOnce    sync.Once
+	initialized bool
+}
+
+func (e *operatorWrapper) Init(params *params.Params) (err error) {
+	e.initOnce.Do(func() {
+		err = e.Operator.Init(params)
+		e.initialized = true
+	})
+	return err
+}
+
+// Register adds a new operator to the registry
+func Register(operator Operator) {
+	if _, ok := allOperators[operator.Name()]; ok {
+		panic(fmt.Errorf("operator already registered: %q", operator.Name()))
+	}
+	allOperators[operator.Name()] = operator
```

Comment:
```suggestion
	allOperators[operator.Name()] = &operatorWrapper{Operator: operator}
```

Otherwise the init once logic won't work.